src/traversals/dfs.jl

```diff
@@ -168,6 +199,37 @@ function topological_sort_by_dfs(g::AbstractGraph)
     reverse(visitor.vertices)
 end
 
+function topological_sort_by_dfs2(g::AbstractGraph)
+    T = eltype(g)
+    vcolor = zeros(UInt8, nv(g))
+    verts = Vector{T}()
+    for v in vertices(g)
+        vcolor[v] != 0 && continue
+        S = Vector{T}([v])
+        vcolor[v] = 1
+        while !isempty(S)
+            u = S[end]
+            w = 0
+            for n in out_neighbors(g, u)
+                if vcolor[n] == 1
+                    error("The input graph contains at least one loop.")
+                elseif vcolor[n] == 0
+                    w = n
+                    break
+                end
+            end
+            if w != 0
+                vcolor[w] = 1
+                push!(S, w)
+            else
+                vcolor[u] = 2
+                push!(verts, u)
+                pop!(S)
+            end
+        end
+    end
+    return reverse(verts)
+end
```

Comment:
I would rather this error now and add functionality later, because of semantic versioning.